### DISCUSSION TITLE
How to trigger required error on string?

### DISCUSSION THREAD

**Discussion by leepowelldev** [colinhacks/zod#1608] [Q&A] [answered]
Given: 
```ts
const name = z.string({ required_error: 'This is required' });
```
How can I trigger a required error on this schema? If I do `name.parse('')` it passes, and if I do `name.parse(undefined)` I get an invalid type error. So what are the requirements for a string being required?

**Comment by JacobWeisenburger** [comment, marked as answer]
## short answer
In the scenario you have laid out, you are technically getting a `required_error`. It is just confusing because the error code for that is `invalid_type`. I totally think this should be changed, but unfortunately I don't have time to put in a PR for this.

You are always welcome to submit a PR to fix this.

## feel free to verify with this code
```ts
const stringSchema = z.string( {
    required_error: 'required_error',
    invalid_type_error: 'invalid_type_error',
} )

const testValues = [
    'string',
    undefined,
    1,
]

testValues.forEach( value => {
    const result = stringSchema.safeParse( value )
    const { success } = result

    if ( result.success ) {
        console.log( { value, success } )
    } else {
        const { code, message } = result.error.issues[ 0 ]
        console.log( { value, code, message } )
    }
} )

/* console output:
{ value: "string", success: true }
{ value: undefined, code: "invalid_type", message: "required_error" }
{ value: 1, code: "invalid_type", message: "invalid_type_error" }
*/
```
